pkg/apis/events/v1alpha1/cloudstoragesource_validation.go

```diff
@@ -73,3 +76,20 @@ func validateSecret(secret *corev1.SecretKeySelector) *apis.FieldError {
 	}
 	return errs
 }
+
+func (current *CloudStorageSource) CheckImmutableFields(ctx context.Context, original *CloudStorageSource) *apis.FieldError {
+	if original == nil {
+		return nil
+	}
+	// Modification of EventType, Secret, PubSubSecret, and Project are not allowed. Everything else is mutable.
+	if diff := cmp.Diff(original.Spec, current.Spec,
+		cmpopts.IgnoreFields(CloudStorageSourceSpec{},
+			"Sink", "CloudEventOverrides", "ServiceAccountName", "Bucket", "ObjectNamePrefix", "PayloadFormat")); diff != "" {
```

Comment:
let's do Bucket, ObjectNamePrefix and PayloadFormat also immutable...
Otherwise, if we change the bucket, we will be creating a new notification in that bucket, but leak the previous one... 
The reconciler should actually delete the previous, and create the new one...

I'd rather stay simple now, and make all of those immutable, and then we can improve the reconciler, and relax this condition... What do you think?

Reply:
BTW, would you mind adding this immutability to Scheduler as well? That can be in a separate PR, whatever works

Reply:
Thanks for the review. Added immutability checks + UTs for all the requested fields.
Will prepare a separate PR to add immutability to Scheduler. Let me know if there is anything else pending here. Thanks again.

Reply:
Created https://github.com/google/knative-gcp/pull/538 for immutability for Scheduler.